### PR TITLE
Optimized lock for per-pool locking

### LINKED ISSUE #2724: Optimize lock for per-pool locking
We've introduced multithreading into stratisd but have no yet managed to optimize a lot for per-pool locking.

### PR DESCRIPTION
**Pull request by jbaublitz**
Follow up to #2788 
Closes #2724

### CODE REVIEW

**Review by mulkieran — CHANGES_REQUESTED** [review]
Also, please grab two additional commits from https://github.com/stratis-storage/stratisd/pull/2873. That's it!

File: src/engine/structures/lock.rs [hunk 1]
```diff
@@ -0,0 +1,970 @@
...
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        &*self.0
+    }
+}
+
+impl<G> Drop for SharedGuard<G> {
+    fn drop(&mut self) {
+        trace!("Dropping shared lock {}", type_name::<G>());
```
Comment (mulkieran) [inline src/engine/structures/lock.rs:43]: Plz insert "on" into for greater consistency with other similar messages, as "Dropping shared lock on {}".

File: src/engine/structures/lock.rs [hunk 2]
```diff
@@ -0,0 +1,970 @@
...
+    G: DerefMut,
+{
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut *self.0
+    }
+}
+
+impl<G> Drop for ExclusiveGuard<G> {
+    fn drop(&mut self) {
+        trace!("Dropping exclusive lock {}", type_name::<G>());
```
Comment (mulkieran) [inline src/engine/structures/lock.rs:72]: Also "on" here.

File: src/engine/structures/lock.rs [hunk 3]
```diff
@@ -0,0 +1,970 @@
...
+#[derive(Debug, PartialEq)]
+pub enum WaitType<U> {
+    SomeRead(U),
+    SomeWrite(U),
+    AllRead,
+    AllWrite,
+}
+
+/// A record of a waiting request.
+pub struct Waiter<U> {
```
Comment (mulkieran) [inline src/engine/structures/lock.rs:415]: Drop "pub" modifier

File: src/engine/structures/lock.rs [hunk 4]
```diff
@@ -0,0 +1,970 @@
...
+            .field("write_locked", &self.write_locked)
+            .field("waiting", &self.waiting)
+            .field("woken", &self.woken)
+            .finish()
+    }
+}
+
+/// A record of the type of a waiting request.
+#[derive(Debug, PartialEq)]
+pub enum WaitType<U> {
```
Comment (mulkieran) [inline src/engine/structures/lock.rs:407]: Drop "pub" modifier.

**Review by mulkieran — CHANGES_REQUESTED** [review]

**Review by mulkieran — APPROVED** [review]

**Comment by jbaublitz** [comment]
Ready for merge.

**Pull request merged**